### PR TITLE
Complex args

### PR DESCRIPTION
**Pull request by marcincuber**
1. Support for complex Parameter Values
2. Support for complex Tag Values
3. Added Unit tests, full coverage
4. Version upgrade

### CODE REVIEW

File: lib/convert.js [hunk 1]
```diff
@@ -3,17 +3,33 @@
...
+      )
+      (?=             FOLLOWED BY
+          \s*,        0 or more empty spaces and a comma
+          |           OR
+          \s*$        0 or more empty spaces and nothing else (end of string)
+      )
+  */
+  
+  param_arr.forEach((pair) => {
+    const split_pair = pair.split(/=(.+)?/);
```
Comment (jackharvey1) [inline lib/convert.js]: `const [key, value] = pair.split(/=(.+)?/);` for readability?

File: lib/convert.js [hunk 2]
```diff
@@ -34,14 +50,16 @@ const opt_string = (value) => {
 
 const opt_tag = (tag) => {
   const retrived_tag = {};
-  const pairs = tag.split(',').map((pair) => pair.split('='));
+  const tag_arr = tag.match(/(\s*([^,]+)".*?"|\s*([^,]+)[^",\s]+)(?=\s*,|\s*$)/g); //same regex as in param_arr
```
Comment (jackharvey1) [inline lib/convert.js]: I wouldn't put this here, this comment's gonna become out of date and then it'll be a lie

File: lib/process_cli_options.js [hunk 3]
```diff
@@ -19,7 +19,7 @@ const processopts = (argv) => {
 
   const argstring = (arg) => {
     const val = argv[arg];
-
+    /* istanbul ignore if */
```
Comment (jackharvey1) [inline lib/process_cli_options.js]: Why are you ignoring this?

File: lib/convert.js [hunk 4]
```diff
@@ -3,17 +3,33 @@
...
-  pairs.forEach((pair) => {
-    if (pair[0] === 'ParameterKey') {
-      retrived_param.ParameterKey = pair[1];
+    if (split_pair[0] === 'ParameterKey') {
+      retrived_param.ParameterKey = split_pair[1];
     } 
-    else if (pair[0] === 'ParameterValue') {
-      retrived_param.ParameterValue = pair[1];
+    else if (split_pair[0] === 'ParameterValue') {
+      retrived_param.ParameterValue = split_pair[1].replace(/^"(.*)"$/, '$1');
```
Comment (jackharvey1) [inline lib/convert.js]: `... .replace('"', '')?`

File: lib/convert.js [hunk 5]
```diff
@@ -3,17 +3,33 @@
...
+          [^",\s]+    1 or more characters excl. double quotes, comma or spaces of any kind
+      )
+      (?=             FOLLOWED BY
+          \s*,        0 or more empty spaces and a comma
+          |           OR
+          \s*$        0 or more empty spaces and nothing else (end of string)
+      )
+  */
+  
+  param_arr.forEach((pair) => {
```
Comment (jackharvey1) [inline lib/convert.js]: Putting all those together:
```
const opt_params = (param) => {
  const param_arr = param.match(/(\s*([^,]+)".*?"|\s*([^,]+)[^",\s]+)(?=\s*,|\s*$)/g);
  
  return param_arr.reduce((retrieved, pair) => {
    const [key, value] = pair.split(/=(.+)?/);
    if (key === 'ParameterKey') {
       retrieved.ParameterKey = value;
     } else if (key === 'ParameterValue') {
       retrieved.ParameterValue = value.replace('"', '');
     } else if (key === 'UsePreviousValue') {
       retrieved.UsePreviousValue = value === 'true';
     } else {
       throw new Error('Missing ParameterKey, ParameterValue or UsePreviousValue');
     }
   });
 
   return retrieved;
 };
}, {});
```

File: lib/convert.js [hunk 6]
```diff
@@ -34,14 +50,16 @@ const opt_string = (value) => {
 
 const opt_tag = (tag) => {
   const retrived_tag = {};
-  const pairs = tag.split(',').map((pair) => pair.split('='));
+  const tag_arr = tag.match(/(\s*([^,]+)".*?"|\s*([^,]+)[^",\s]+)(?=\s*,|\s*$)/g); //same regex as in param_arr
 
-  pairs.forEach((pair) => {
-    if (pair[0] === 'Key') {
-      retrived_tag.Key = pair[1];
+  tag_arr.forEach((tag) => {
```
Comment (jackharvey1) [inline lib/convert.js]: Similarly

```
const opt_tag = tag => {
  const tag_arr = tag.match(/(\s*([^,]+)".*?"|\s*([^,]+)[^",\s]+)(?=\s*,|\s*$)/g); //same regex as in param_arr

  tag_arr.redtuce((retrieved, tag) => {
    const [key, value] = tag.split(/=(.+)?/);

    if (split_tag[0] === 'Key') {
      retrieved.Key = split_tag[1];
    } else if (split_tag[0] === 'Value') {
      retrieved.Value = split_tag[1].replace(/^"(.*)"$/, '$1');
     } else {
       throw new Error('Missing Key or Value inside tag');
     }
   });
 
   return retrived_tag;
 };
```

File: lib/convert.js [hunk 7]
```diff
@@ -34,14 +50,16 @@ const opt_string = (value) => {
...
+    const split_tag = tag.split(/=(.+)?/);
+
+    if (split_tag[0] === 'Key') {
+      retrived_tag.Key = split_tag[1];
     } 
-    else if (pair[0] === 'Value') {
-      retrived_tag.Value = pair[1];
+    else if (split_tag[0] === 'Value') {
```
Comment (jackharvey1) [inline lib/convert.js]: Is there a chance these could be the same? Would running the `replace` on a key be fatal? In which case you can just do (using same variables as my reduce above)

```
    if (key === 'Key' || key === 'Value') {
      retrieved[key] = value.replace(/^"(.*)"$/, '$1');
    } else {
      throw new Error('Missing Key or Value inside tag');
    }
```

File: lib/convert.js (hunk 4 above)
Comment (marcincubernews) [reply lib/convert.js]: I believe `.replace('"', '')` is removing all the double quotes whereas `replace(/^"(.*)"$/, '$1')` is removing only the first and last `"`.

File: .travis.yml [hunk 8]
```diff
@@ -10,3 +10,4 @@ script:
   - npm run test
   - npm run coverage
   - npm run report-coverage
+  - codecov
```
Comment (Belema) [inline .travis.yml]: Subjective

File: lib/convert.js (hunk 4 above)
Comment (jackharvey1) [reply lib/convert.js]: `.replace(/(^"|"$)/g, '')`?

File: .travis.yml (hunk 8 above)
Comment (marcincubernews) [reply .travis.yml]: Thanks for comprehensive review 👍

**Pull request merged**